gflownet/gflownet.py

```diff
@@ -591,43 +643,143 @@ def trajectorybalance_loss(self, it, batch):
+                "data must be either a list of states or a path to a .pkl file."
+            )
+        assert len(states_term) < max_data
+        envs = []
+        for state_idx, x in enumerate(states_term):
+            for traj_idx in range(n_trajectories):
+                idx = int(max_data * state_idx + traj_idx)
+                env = self.env.copy().reset(idx)
+                env.set_state(x, done=True)
+                envs.append(env)
+        # Sample trajectories
+        actions = []
+        valids = []
+        max_iters = n_trajectories * max_iters_per_traj
+        iters = 0
+        while envs:
+            # Sample backward actions
+            actions = self.sample_actions(
+                envs,
+                batch,
+                backward=True,
+                no_random=True,
+                times=times,
+            )
+            # Update environments with sampled actions
+            envs, actions, valids = self.step(envs, actions, backward=True)
+            # Add to batch
+            batch.add_to_batch(envs, actions, valids, backward=True, train=True)
+            assert all(valids)
+            # Filter out finished trajectories
+            envs = [env for env in envs if not env.equal(env.state, env.source)]
+        # Prepare data structures to compute log probabilities
+        traj_ids = np.array([k for k in batch.trajectories])
+        data_indices = tlong(traj_ids // max_data, device=self.device)
+        traj_indices = tlong(traj_ids % max_data, device=self.device)
+        logprobs_f = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -torch.inf,
             dtype=self.float,
             device=self.device,
-        ).index_add_(0, traj_indices, logprobs_f)
-        # Backward trajectories
-        policy_output_b = self.backward_policy(states)
-        logprobs_b = self.env.get_logprobs(
-            policy_output_b, False, actions, parents, masks_b
         )
-        sumlogprobs_b = torch.zeros(
-            batch.get_n_trajectories(),
+        logprobs_b = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -torch.inf,
             dtype=self.float,
             device=self.device,
-        ).index_add_(0, traj_indices, logprobs_b)
-        # Trajectory balance loss
-        loss = (
-            (self.logZ.sum() + sumlogprobs_f - sumlogprobs_b - torch.log(rewards))
-            .pow(2)
-            .mean()
         )
-        return loss, loss, loss
+        traj_indices_mat = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -1,
+            dtype=torch.long,
+            device=self.device,
+        )
+        traj_indices_mat[data_indices, traj_indices] = traj_indices
+        n_trajs_per_sample, _ = traj_indices_mat.max(dim=1)
+        # Compute log probabilities of the trajectories
+        logprobs_f[data_indices, traj_indices] = self.compute_logprobs_trajectories(
```

Comment:
it seems to me that this line (and the next one) relies on the fact that environments were added to the batch in a specific order, i.e. [first env for the first data sample, second env for the first data sample, ...]. There's nothing wrong with it and this order seems to be satisfied in the loop above, but maybe it's worth mentioning in a comment that the order is very important (otherwise we will mess up probs for different data samples)

Reply:
The order is ensured by both [this line](https://github.com/alexhernandezgarcia/gflownet/pull/167/files#diff-621df8cda7957075a15e68cd043eec610a0e71cb65791028ad4092ac1b725bbaR538) and because the ids of the environments are set in a specific way [here](https://github.com/alexhernandezgarcia/gflownet/pull/167/files#diff-621df8cda7957075a15e68cd043eec610a0e71cb65791028ad4092ac1b725bbaR729).

Reply:
I have added a note in the docstring anyway [here](https://github.com/alexhernandezgarcia/gflownet/pull/167/commits/af8043c85175ada437ea05c9dcc030372bf407c6).

Reply:
Made clearer in https://github.com/alexhernandezgarcia/gflownet/pull/167/commits/3599dbf13a7cf0e0bead4b937a85a9818a215cd9.

Reply:
Actually, it doesn't rely on that because of the indexing provided by `data_indices` and `traj_indices`. I hope it is clearer now.